strawberryfields/apps/train/parametrization.py

```diff
@@ -0,0 +1,406 @@
+               [0.30298161, 0.        , 0.30756059, 0.30910225],
+               [0.31534653, 0.30756059, 0.        , 0.32171695],
+               [0.31692721, 0.30910225, 0.32171695, 0.        ]])
+
+        Args:
+            params (array): the trainable parameters :math:`\theta`
+
+        Returns:
+            array: the trained adjacency matrix
+        """
+        return self.W(params) @ self.A_init @ self.W(params)
+
+    def generate_samples(self, A: np.ndarray, n_samples: int, **kwargs) -> np.ndarray:
+        """Generate GBS samples from an input adjacency matrix.
+
+        **Example usage:**
+
+        >>> vgbs.generate_samples(vgbs.A(params), 2)
+        array([[1, 1, 1, 0],
+               [0, 1, 0, 1]])
+
+        Args:
+            A (array): the adjacency matrix
+            n_samples (int): the number of GBS samples to generate
+            **kwargs: additional arguments to pass to the sampler from
+                `The Walrus <https://the-walrus.readthedocs.io/en/stable/>`__
+
+        Returns:
+            array:  the generated samples
+        """
+        cov = A_to_cov(A)
+
+        if self.threshold:
+            samples = thewalrus.samples.torontonian_sample_state(
+                cov, n_samples, hbar=sf.hbar, **kwargs
+            )
+        else:
+            samples = thewalrus.samples.hafnian_sample_state(cov, n_samples, hbar=sf.hbar, **kwargs)
+        return samples
+
+    def add_A_init_samples(self, samples: np.ndarray):
+        r"""Add samples of the initial adjacency matrix to :attr:`A_init_samples`.
+
+        .. warning::
+
+            The added samples must be from the *input* adjacency matrix and not the trained one
+            :math:`A(\theta)`.
+
+        **Example usage:**
+
+        >>> samples = np.array([[0, 1, 0, 0], [0, 1, 1, 1]])
+        >>> vgbs.add_A_init_samples(samples)
+
+        Args:
+            samples (array): samples from the initial adjacency matrix
+        """
+        if samples is None:
+            return
+
+        shape = samples.shape
+        if shape[1] != self.n_modes:
+            raise ValueError("Must input samples of shape (number, {})".format(self.n_modes))
+
+        if self.A_init_samples is None:
+            self.A_init_samples = samples
+        else:
+            self.A_init_samples = np.vstack([self.A_init_samples, samples])
+
+    def get_A_init_samples(self, n_samples: int) -> np.ndarray:
+        """Get samples from the initial adjacency matrix.
+
+        This function checks for pre-generated samples in ``A_init_samples``. If there are fewer
+        than ``n_samples`` stored, more samples are generated and added to ``A_init_samples``.
+
+        Args:
+            n_samples (int): number of samples to get
+
+        Returns:
+            array: samples from the initial adjacency matrix
+        """
+        try:
+            current_n_samples = self.A_init_samples.shape[0]
+        except AttributeError:
+            current_n_samples = 0
```

Comment:
Curious, which cases does this cover?
a) `self.A_init_samples is None`
b) `not isinstance(self.A_init_samples, np.ndarray)`
c) there or something else too

If it's either a) or b) perhaps it's good to go with the explicit case. If it's c), then curious what I've missed :smile:

Reply:
Ohh yes this is a good question for code quality.

So yeah generally I expect `A_init_samples` to be either `None` or `np.ndarray`. I want to check the shape, and of course if `A_init_samples = None` then assume there are no samples present.

One options is `isinstance(A_init_samples, np.ndarray)`, but this requires `A_init_samples` to be exactly an `np.ndarray`.

Another option I guess is to check if `A_init_samples` is None :thinking: 

But then I thought to follow the EAFP principle and just try and query the shape.

So I guess this option allows for the rare occasion that a user puts in something like a numpy array but not exactly (e.g. if they subclass it).

But interested to hear your thoughts!